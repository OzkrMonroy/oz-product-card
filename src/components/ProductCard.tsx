import React, { createContext } from 'react';
import styles from '../styles/styles.module.css';
import {
  ProductCardProps,
  ProductContextProps,
} from '../interfaces/productInterfaces';
import { useProduct } from '../hooks/useProduct';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export const ProductCard = ({
  product,
  children,
  className,
  style,
  onChange,
  count,
  initialValues,
}: ProductCardProps) => {
  const { counter, increaseBy, isMaxCountReached, reset } = useProduct({
    product,
    onChange,
    initialCounter: count,
    initialValues,
  });

  return (
    <Provider
      value={{
        product,
        counter,
        increaseBy,
        maxCount: initialValues?.maxCount,
      }}
    >
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          count: counter,
          isMaxCountReached,
          reset,
          maxCount: initialValues?.maxCount,
          product,
          increaseBy,
        })}
        {/* <ProductImage img={img} name={name} />
      <ProductTitle name={name} />
      <ProductsButtons counter={counter} increaseBy={increaseBy}/> */}
      </div>
    </Provider>
  );
};
