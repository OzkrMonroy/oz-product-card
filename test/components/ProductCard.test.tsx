import React from 'react';
import renderer from 'react-test-renderer';
import ProductCard from '../../src/components/index';
import { product2 } from '../mocks/products';

const { act } = renderer;

describe('ProductCard component', () => {
  test('should render the component correctly', () => {
    const wrapper = renderer.create(
      <ProductCard product={product2}>{() => <h1>Hello world</h1>}</ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('should increment the counter', () => {
    const wrapper = renderer.create(
      <ProductCard product={product2}>
        {({ increaseBy, count }) => (
          <>
            <h1>Hello world</h1>
            <span>{count}</span>
            <button onClick={() => increaseBy(1)}></button>
          </>
        )}
      </ProductCard>
    );

    let tree = wrapper.toJSON();
    expect(tree).toMatchSnapshot();

    act(() => {
      (tree as any).children[2].props.onClick();
    });

    tree = wrapper.toJSON();

    expect((tree as any).children[1].children[0]).toBe('1');
  });
});
