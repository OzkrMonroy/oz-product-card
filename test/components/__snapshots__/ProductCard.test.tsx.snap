// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard component should increment the counter 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Hello world
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  />
</div>
`;

exports[`ProductCard component should render the component correctly 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Hello world
  </h1>
</div>
`;
